File: 04day/part1.py
def solve(line: str) -> int:
    first, second = line.split(",")
    f1, f2 = first.split("-")
    s1, s2 = second.split("-")
    f1, f2, s1, s2 = int(f1), int(f2), int(s1), int(s2)
    if (f1 <= s1 and f2 >= s2) or (s1 <= f1 and s2 >= f2):
        return 1
    return 0


if __name__ == "__main__":
    lines = []
    with open("input.txt") as f:
        for line in f:
            lines.append(line)
    print(sum(map(solve, lines)))

File: 04day/part1.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.Text (pack, split, unpack)
import System.IO (IOMode (ReadMode), hClose, hGetContents, openFile)

solve :: String -> Int
solve xs
  | f1 <= s1 && f2 >= s2 || s1 <= f1 && s2 >= f2 = 1
  | otherwise = 0
  where
    splits = split (== '-') (pack xs)
    beginEnds = concatMap (split (== ',')) splits
    [f1, f2, s1, s2] = map ((read :: String -> Int) . unpack) beginEnds

solution :: String -> Int
solution = sum . map solve . lines

main :: IO ()
main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print $ solution contents
  hClose handle